src/cdk-experimental/menu/menu-stack.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Subject, Observable} from 'rxjs';
+
+/** Events to emit as specified by the caller once the MenuStack is empty. */
+export type FocusNext = 'nextItem' | 'previousItem' | 'currentItem' | undefined;
+
+/**
+ * Interface for the elements tracked in the MenuStack.
+ */
+export interface MenuStackItem {
+  /** A reference to the previous Menus MenuStack instance. */
+  _menuStack: MenuStack;
+}
+
+/**
+ * MenuStack allows subscribers to listen for close events (when a MenuStackItem is popped off
+ * of the stack) in order to perform closing actions. Upon the MenuStack being empty it emits
+ * from the `empty` observable specifying the next focus action which the listener should perform
+ * as requested by the closer.
+ */
+export class MenuStack {
+  /** All MenuStackItems tracked by this MenuStack. */
+  private readonly _elements: MenuStackItem[] = [];
```

Comment:
I think that using a stack/array that pretends to be a stack here will run into the same issue as when we tried to keep track of menu items ourselves in `mat-menu`. If a menu pushes itself onto the stack, gets destroyed by an `ngIf` and then gets re-added, the stack will be out of order. E.g. if we have `[1, 2, 3]`, we remove 2 and then we re-add it, the stack will have `[1, 3, 2]`, whereas the DOM will still be `[1, 2, 3]`.

That being said, I'm having a hard time visualizing when the registration happens. If it's only while a stack of menus is open, we'll probably be fine.

Reply:
Hmm. Not sure if I follow. If we remove 2 that means we closed 2 and closing 2 should close all submenus. Hence `[1,2,3]` would result in `[1]` after removing 2.

Reply:
Yeah, I _think_ that if you ensure any menu closing also closes all of its sub-menus, this should be okay. (Unless there's something a user might do that I'm not imagining)

Reply:
Was already merged in before I had a chance to address this. I'll address with a follow on pr.